File: contract/src/main/java/rrlane/leasing/contract/service/VehicleServiceImpl.java
package rrlane.leasing.contract.service;

import rrlane.leasing.contract.entity.Vehicle;
import rrlane.leasing.contract.repo.VehicleRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
public class VehicleServiceImpl implements VehicleService {
    @Autowired
    private VehicleRepository vehicleRepository;

    @Override
    @Transactional
    public void saveVehicle(Vehicle vehicle) {
        System.out.println("Saving vehicle...");
        vehicleRepository.save(vehicle);
        System.out.println("Saved vehicle successfully.");
    }

    public List<Vehicle> getVehiclesByDetails(String[] details){
        List<Vehicle> foundVehicles = null;
        foundVehicles = vehicleRepository.findByBrand(details[0]);
        if(null == foundVehicles)
        foundVehicles = vehicleRepository.findByModel(details[1]);
        if(null == foundVehicles)
        foundVehicles = vehicleRepository.findByModelYear(details[2]);
        if(null == foundVehicles)
        foundVehicles = vehicleRepository.findByVin(details[3]);
        return foundVehicles;
    }
}

File: contractapi/src/main/kotlin/de/leasing/contract/api/service/VehicleServiceImpl.kt
package de.leasing.contract.api.service

import de.leasing.contract.api.entity.Vehicle
import de.leasing.contract.api.entity.dto.VehicleReq
import de.leasing.contract.api.entity.dto.VehicleResp
import de.leasing.contract.api.exception.ResourceNotFoundException
import de.leasing.contract.api.repo.VehicleRepository
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service

@Service
class VehicleServiceImpl(private val vehicleRepository: VehicleRepository) :
    VehicleService {

    private val log: Logger = LoggerFactory.getLogger(VehicleServiceImpl::class.java)

    override fun saveVehicleRecord(vehicleReq: VehicleReq): VehicleResp {
        log.info("Saving the details of vehicle...")
        return vehicleRepository.save(Vehicle.fromReq(vehicleReq)).toResp()
    }

    override fun listAllVehicleRecords(page: Int, size: Int): Iterable<VehicleResp> {
        val pageable: Pageable = PageRequest.of(page, size)
        val vehicleResps = vehicleRepository.findAll(pageable).map { it.toResp() }
        if (vehicleResps.isEmpty()) throw ResourceNotFoundException("No records of vehicles found.")
        for(v in vehicleResps){
            print("vehicleResps: $v")}
        return vehicleResps
    }

    override fun listAllVehicleRecords(sortBy: String?): Iterable<VehicleResp> {
        val sortedVehicleResps = if (null == sortBy) vehicleRepository.findAll()
            .map { it.toResp() } else vehicleRepository.findAll(
            Sort.by(
                Sort.Direction.DESC, sortBy
            )
        ).map { it.toResp() }
        if (sortedVehicleResps.isEmpty()) throw ResourceNotFoundException("No vehicle records available to sort and present.")
        return sortedVehicleResps
    }

    override fun getVehicleById(id: String): VehicleResp {
        return vehicleRepository.getReferenceById(id).toResp()
    }

}